### PR TITLE
[Messenger] Add a trait for synchronous query & command buses

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | 4.2 
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | N/A   
| License       | MIT
| Doc PR        | symfony/symfony-docs/issues/10662

### CODE REVIEW

File: src/Symfony/Component/Messenger/QueryBusInterface.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace Symfony\Component\Messenger;
+
+/**
+ * Opinionated version of a message bus, expecting a single, synchronous message handling and returning its result.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.3
+ */
+interface QueryBusInterface extends MessageBusInterface
```
[the remark this answers is not in the document]
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/QueryBusInterface.php]: Alright, removed for now.

File: src/Symfony/Component/Messenger/QueryBus.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+namespace Symfony\Component\Messenger;
+
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+
+/**
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.3
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/QueryBus.php]: I'd be for merging this on 4.2 so that we can make the component stable for 4.3

**Comment by Tobion** [comment]
While the implementation looks clean and simple, I'm not confident that a query bus should be added at all. The problem that @nicolas-grekas already mentioned in https://github.com/symfony/symfony/pull/28909#issuecomment-430889273 still applies. This does not allow to use type declarations for the return value. And there are better alternatives already.

Also I would argue that queries should not be using the command bus pattern at all. Use the repository pattern for this. This is also why SimpleBus does not include a query bus.

**Comment by Tobion** [comment]
This PR replaces #28716 I assume

File: src/Symfony/Component/Messenger/QueryBus.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function query($message)
+    {
+        $envelope = $this->messageBus->dispatch($message);
+        /** @var HandledStamp[] $handledStamps */
+        $handledStamps = $envelope->all(HandledStamp::class);
+
+        if (!isset($handledStamps[0])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBus.php]: Should be if (!$handledStamps) to work with null values.

File: src/Symfony/Component/Messenger/QueryBus.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            throw new LogicException(sprintf(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBus.php]: We always try to put exceptions on one line.

File: src/Symfony/Component/Messenger/QueryBus.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                'Query of type "%s" wasn\'t handled. Is the "%s" middleware properly registered in your bus?',
+                \get_class($envelope->getMessage()),
+                HandleMessageMiddleware::class
+            ));
+        }
+
+        if (isset($handledStamps[1])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBus.php]: Should use count() to deal with null

**Review by nicolas-grekas — COMMENTED** [review]
What about shipping a trait instead of an interface + class? We would then encourage ppl to declare return types when using it. The method added to the trait would be private.

**Comment by ogizanagi** [comment]
@nicolas-grekas 's still applies, yes....about autocompletion. While I would usually agree with you, here the return type wouldn't even be of any help in most cases, because the result is often simply sent to the template or a serializer.

> This does not allow to use type declarations for the return value

Return type is enforced by the handler. Which we ensure is registered, and single. So a mismatch in config/code may still happen, but at some point users are responsible of their config and the issue would be easily spotted.

> Also I would argue that queries should not be using the command bus pattern at all. Use the repository pattern for this.

The query bus isn't a replacement of the repository pattern but is actually complementary to it.
It also allows:
- cross-cutting concerns through middleware (logging, cache, storing metrics, query validation, security, ...)
- centralizing application use-cases in one place, in a consistent way between reads & writes. It's awesome for discoverability, really helpful for anyone new in the team to understand the application.

Query handlers are also the place where data is aggregated into DTOs matching the use-case.
All of this favor best practices & designing sustainable, understandable, enjoyable apps.

If one really needs the return type, this is perfectly legit to implement in its action or any caller:

```php
public function query($query): MyQueryResult
{
    return $this->queryBus->query($query);
}
```

And if we want to enforce return type to prevent any misconfig issue, we could add an optional fqcn argument to assert the result is what we expect.

> What about shipping a trait instead of an interface + class? We would then encourage ppl to declare return types when using it. The method added to the trait would be private.

This is not needed to me regarding above explanations.

**Comment by nicolas-grekas** [comment]
I insist in the trait.
What is not needed is the interface and the class ;)

**Comment by ogizanagi** [comment]
Trait it is...

(Agree for the interface, but I would have provided the base class anyway. Expect to see this happen in  userland.)

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+trait QueryBusTrait
+{
+    /**
+     * Dispatches the given message and returns the result from the handler returned value.
+     *
+     * @param object|Envelope $message The message or the message pre-wrapped in an envelope
+     *
+     * @return mixed The handler returned value
+     */
+    public function dispatchQuery(MessageBusInterface $messageBus, $message)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: private function dispatchQuery($message) + private $queryBus property that the consumer would need to set in their constructor (eg type-hinting `MessageBusInterface $queryBus`)

File: src/Symfony/Component/Messenger/Tests/QueryBusTest.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    use QueryBusTrait;
+
+    private $messageBus;
+
+    public function __construct(MessageBusInterface $messageBus)
+    {
+        $this->messageBus = $messageBus;
+    }
+
+    public function query($query)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Tests/QueryBusTest.php]: for the example, i'd suggest adding a return type here

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+     * Dispatches the given message and returns the result from the handler returned value.
+     *
+     * @param object|Envelope $query The message or the message pre-wrapped in an envelope
+     *
+     * @return mixed The handler returned value
+     */
+    private function dispatchQuery($query)
+    {
+        if (!$this->queryBus instanceof MessageBusInterface) {
+            throw new LogicException(sprintf('You must provide a "%s" instance in the "%s::$queryBus" property. "%s" given.', MessageBusInterface::class, \get_class($this), \is_object($this->queryBus) ? \get_class($this->queryBus) : \gettype($this->queryBus)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: first dot should be a comma (single sentence)

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        $envelope = $this->queryBus->dispatch($query);
+        /** @var HandledStamp[] $handledStamps */
+        $handledStamps = $envelope->all(HandledStamp::class);
+
+        if (!isset($handledStamps[0])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: should use "count" to account for nulls

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            throw new LogicException(sprintf(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: should be on one line

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                'Query of type "%s" wasn\'t handled. Is the "%s" middleware properly registered in your bus?',
+                \get_class($envelope->getMessage()),
+                HandleMessageMiddleware::class
+            ));
+        }
+
+        if (isset($handledStamps[1])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: should use "count" to account for nulls

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            $handlers = implode(', ', array_map(function (HandledStamp $stamp): string {
+                return sprintf('"%s"', $stamp->getHandlerAlias() ?? $stamp->getCallableName());
+            }, $handledStamps));
+            throw new LogicException(sprintf('Query of type "%s" was handled multiple times (responding handlers: %s). Check your configuration or the debug command to ensure there is only one possible handler for this query.', \get_class($envelope->getMessage()), $handlers));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: Not sure about the 2nd sentence: the command is not needed when using the component standalone. I think we can just remove it.

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\Middleware\HandleMessageMiddleware;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+
+/**
+ * Leverages a message bus to expect a single, synchronous message handling and return its result.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.3
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: 4.2?

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 14]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+4.3.0
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/CHANGELOG.md]: 4.2?

**Review by nicolas-grekas — COMMENTED** [review]
> I would have provided the base class anyway

a class is still a type, which is what is costly in terms of maintenance / SOLID. A trait is a pure behavior so mostly free from such considerations.

**Comment by Tobion** [comment]
I agree the interface is not needed. But the query bus as a class is much cleaner than as a trait. Traits are for code reuse in single-inheritance languages. A class like a controller that requires a query bus, depends on it and it does not behave like one.

File: src/Symfony/Component/Messenger/QueryBusTrait.php (hunk 9 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/QueryBusTrait.php]: Hmm. Actually not. Even if `null` is returned by the handler, `$handledStamps[0]` exists as a `HandledStamp` instance.
So better revert to `isset`?

File: src/Symfony/Component/Messenger/QueryBusTrait.php (hunk 12 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/QueryBusTrait.php]: I've removed it, but this is actually quite a generic hint. At least if you use the component with the fwb, you might learn this command exists.

File: src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        );
+
+        $trait->dispatchQuery($query);
+    }
+}
+
+class TestQueryBusTrait
+{
+    use QueryBusTrait {
+        dispatchQuery as public;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php]: I'd suggest the same as before: keep the method private and add a "public query(): string" ;)

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+        if (!$handledStamps) {
+            throw new LogicException(sprintf('Query of type "%s" wasn\'t handled. Is the "%s" middleware properly registered in your bus?', \get_class($envelope->getMessage()), HandleMessageMiddleware::class));
+        }
+
+        if (\count($handledStamps) > 1) {
+            $handlers = implode(', ', array_map(function (HandledStamp $stamp): string {
+                return sprintf('"%s"', $stamp->getHandlerAlias() ?? $stamp->getCallableName());
+            }, $handledStamps));
+            throw new LogicException(sprintf('Query of type "%s" was handled multiple times (responding handlers: %s).', \get_class($envelope->getMessage()), $handlers));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: What about that?
Message of type "%s" was handled multiple times. Only one handler is expected when using a query bus (responding handlers: %s).

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        if (!$this->queryBus instanceof MessageBusInterface) {
+            throw new LogicException(sprintf('You must provide a "%s" instance in the "%s::$queryBus" property, "%s" given.', MessageBusInterface::class, \get_class($this), \is_object($this->queryBus) ? \get_class($this->queryBus) : \gettype($this->queryBus)));
+        }
+
+        $envelope = $this->queryBus->dispatch($query);
+        /** @var HandledStamp[] $handledStamps */
+        $handledStamps = $envelope->all(HandledStamp::class);
+
+        if (!$handledStamps) {
+            throw new LogicException(sprintf('Query of type "%s" wasn\'t handled. Is the "%s" middleware properly registered in your bus?', \get_class($envelope->getMessage()), HandleMessageMiddleware::class));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: Message of type "%s" was handled zero times. Exactly one handler is expected when using a query bus. (HandleMessageMiddleware is an optional implementation)

File: src/Symfony/Component/Messenger/QueryBusTrait.php (hunk 16 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/QueryBusTrait.php]: Changed, thanks

File: src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php (hunk 15 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php]: That's for tests and we just want to test `QueryBusTrait ::dispatchQuery`. Everything regarding the return type is performed by assertions.
I don't see the need to showcase faked semantics on a `QueryBusTraitConsumer` which is neither a query bus (fixed return type), neither a real trait user (the method won't be called publicly in real use-cases).

Comment (ogizanagi) [reply src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php]: I explained above: a QueryBus implementation won't have a fixed return type (here, `string` as you asked). Also this are tests. We use assertions on tested methods returned values.

But fine I'll changed it back.

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php]: >  a QueryBus implementation won't have a fixed return type

I don't agree with that: it's perfectly fine and better to create typed query buses! I would recommend *never* creating untyped query buses actually.

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Messenger/QueryBusTrait.php [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (!$handledStamps) {
+            throw new LogicException(sprintf('Message of type "%s" was handled zero times. Exactly one handler is expected when using a query bus.', \get_class($envelope->getMessage())));
+        }
+
+        if (\count($handledStamps) > 1) {
+            $handlers = implode(', ', array_map(function (HandledStamp $stamp): string {
+                return sprintf('"%s"', $stamp->getHandlerAlias() ?? $stamp->getCallableName());
+            }, $handledStamps));
+            throw new LogicException(sprintf('Message of type "%s" was handled multiple times. Only one handler is expected when using a query bus, %d did: %s.', \get_class($envelope->getMessage()), \count($handledStamps), $handlers));
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/QueryBusTrait.php]: reads a bit weird, i'd say:

```
Only one handler is expected when using a query bus, got %d: %s.
```

Also newline above the throw? :angel:

**Comment by nicolas-grekas** [comment]
What about renaming this CommandBusTrait?

**Comment by ro0NL** [comment]
>  a message bus instance to return a single synchronous message handling result

That does not a apply to a event- and/or command-bus. Only a query bus,

**Comment by nicolas-grekas** [comment]
In #28716, @pamil uses the term command bus for his need. What's a command bus vs a query bus?

**Comment by nicolas-grekas** [comment]
Or maybe something more neutral like `OneHandlerTrait`?

**Comment by ro0NL** [comment]
I think there 2 concepts to grasp; 
- enforce a single handler
- enforce a result value

The first does not enforce the latter, however both apply to a query bus at least.

**Comment by ogizanagi** [comment]
@ro0NL : We do not enforce a result value. We enforce there is a handler. Which applies to both synchronous command & query buses. Using this trait for a command bus would be fine. Either you choose to use a return value or not, then, is up to you.

**Comment by ro0NL** [comment]
> We do not enforce a result value

True, my bad.

>  suggestion is more about the generic term referring to the command pattern. Not command/query bus

i find that confusing :)

I think something more generic could apply yes, e.g. i have something called `MessageDispatchingTrait`: https://github.com/msgphp/msgphp/blob/master/src/Domain/Message/MessageDispatchingTrait.php#L12

It could provide multiple APIs, e.g. `dispatch()` + `dispatchOne()`

**Review by sroze — APPROVED** [review]
Definitely 👍. Looks great to me.

**Comment by sroze** [comment]
The name `QueryBusTrait` seems reasonable to me as the query bus is typically what describes a bus returning a result, so it sounds good.

If we really want to get rid of the "query bus" notion, then we should focus on the expected behaviour: returning results: `DispatchAndGetResultTrait`?

**Comment by nicolas-grekas** [comment]
Naming things :)
My best proposal would be using just `HandleTrait` + `handle()`:
- use `dispatch()` to get an envelop back and have things potentially async
- use `handle()` from the trait to get synchronous behavior and a result back

**Review by ro0NL — APPROVED** [review]

**Comment by gquemener** [comment]
Sorry to dig up the issue so long after the discussion, but I feel that a single handler is a common strategy when using a command bus.

I've been using `prooph/service-bus` (currently deprecated) for the past years and that's something they provide (https://github.com/prooph/service-bus/blob/master/src/Plugin/Router/SingleHandlerRouter.php).

I'm surprised that the only way the Messenger component offers this feature is through a trait that allows to check that the message was not handled by more than one handler. It feels that this is too late and the damages are already done.

I'd like to work on a PR that tweaks the `HandleMessageMiddleware` to detect this problem before any handler is called (keeping it BC at the same time).

I've read the discussions and couldn't find any moment where this option was discussed. Please forgive me if it is the case.

**Comment by gquemener** [comment]
The idea I have in mind is to add a `allowMultipleHandlers` flag to the middleware, and check how many handlers are located for a given message whenever it's enabled.

**Comment by gquemener** [comment]
Another solution would be to configure the middleware with another (new) implementation of `HandlersLocatorInterface` that throws exception when more than 1 handler is located.

**Comment by gquemener** [comment]
Another option would be to do it at compilation time in the MessengerPass :thinking:

**Comment by chalasr** [comment]
@gquemener Please consider opening a new issue to propose your idea and discuss about it.

**Comment by gquemener** [comment]
@chalasr here it is. Thanks for your reply!

**Pull request merged**